docs/developer/explanations/decisions/0004-repository-structure.rst

```diff
@@ -55,24 +55,40 @@ During this process, the folder structure should incrementally be changed to
     │       │   ├── __init__.py
     │       │   ├── backends
     │       │   │   ├── __init__.py
-    │       │   │   ├── _aioca.py
-    │       │   │   └── _p4p.py
+    │       │   │   ├── signal_backend.py
+    │       │   │   └── sim.py
     │       │   ├── devices
-    │       │   ├── signals
+    │       │   │   ├── __init__.py
+    │       │   │   ├── device_collector.py
```

Comment:
Please can we make everything under `core` start with an underscore (e.g. `backends` -> `_backends`), so the top level import namespace becomes `from ophyd_async.core import SignalRW, DeviceCollector, ...`

Reply:
I'm not sure I like this, though; [PEP8](https://peps.python.org/pep-0008/#public-and-internal-interfaces) suggests the following:

> To better support introspection, modules should explicitly declare the names in their public API using the __all__ attribute. Setting __all__ to an empty list indicates that the module has no public API.

> Even with __all__ set appropriately, internal interfaces (packages, modules, classes, functions, attributes or other names) should still be prefixed with a single leading underscore.

My takeaway from this is; in the __init__.py for `core` we should define the public API. Any objects/functions/modules not exposed in the `__all__` should be prefixed with `_` to mark them as internal. But what you seem to be suggesting here is, we define our public API in `__all__` but then every object that is being exposed publically actually originates from a module file with an underscore. So we are tricking the reader of our code to think that all the `core` code is internal. I think this is a trick; at least as it currently works, the `DeviceCollector` for example will never be an internal tool, and will have to be used for Devices or classes that inherit from Device for certain functionality.

Reply:
It doesn't specifically mention this, but it does imply it

> Wildcard imports (from <module> import *) should be avoided, as they make it unclear which names are present in the namespace, confusing both readers and many automated tools. There is one defensible use case for a wildcard import, which is to republish an internal interface as part of a public API (for example, overwriting a pure Python implementation of an interface with the definitions from an optional accelerator module and exactly which definitions will be overwritten isn’t known in advance).

> When republishing names this way, the guidelines below regarding public and internal interfaces still apply.

I would argue that we are "republish an internal interface as part of a public API", because we want the retain the ability to move functions around private modules, while keeping the same `from ophyd_async.core import DeviceCollector` API.